tor/controller.go

```diff
@@ -259,33 +270,52 @@ func (c *Controller) authenticate() error {
 	return nil
 }
 
-// getAuthCookie retrieves the authentication cookie in bytes from the Tor
-// server. Cookie authentication must be enabled for this to work. The boolean
-func (c *Controller) getAuthCookie() ([]byte, error) {
-	// Retrieve the authentication methods currently supported by the Tor
-	// server.
-	authMethods, cookieFilePath, version, err := c.ProtocolInfo()
+// Contains tells whether a contains x.
+func Contains(a []string, x string) bool {
+	for _, n := range a {
+		if x == n {
+			return true
+		}
+	}
+	return false
+}
+
+// authenticate authenticates the connection between the controller and the
+// Tor server using the SAFECOOKIE authentication method.
+func (c *Controller) authenticate() error {
+	authMethods, version, info, err := c.ProtocolInfo()
 	if err != nil {
-		return nil, err
+		return err
 	}
 
 	// With the version retrieved, we'll cache it now in case it needs to be
 	// used later on.
 	c.version = version
 
-	// Ensure that the Tor server supports the SAFECOOKIE authentication
-	// method.
-	safeCookieSupport := false
-	for _, authMethod := range authMethods {
-		if authMethod == "SAFECOOKIE" {
-			safeCookieSupport = true
-		}
+	// If we have a password, and the HASHEDPASSWORD auth method is
+	// supported, then try that.
+	if Contains(authMethods, "HASHEDPASSWORD") && c.password != "" {
+		return c.authenticateViaHashedPassword()
 	}
 
-	if !safeCookieSupport {
-		return nil, errors.New("the Tor server is currently not " +
-			"configured for cookie authentication")
+	// If the SAFECOOKIE method is supported, try that.
+	if Contains(authMethods, "SAFECOOKIE") {
+		return c.authenticateViaSafeCookie(info)
+	}
+
+	// Don't know how to authenticate. Fail.
+	return fmt.Errorf("[TOR] authenticate: exhausted all auth methods")
```

Comment:
Would prefer this error message to be: `the Tor server must be configured with SAFECOOKIE or HASHEDPASSWORD authentication`.